src/main/java/eu/europeana/metis/sandbox/service/dataset/DatasetGeneratorServiceImpl.java

```diff
@@ -4,52 +4,70 @@
-    requireNonNull(language, "Language must not be null");
+  public Dataset generate(DatasetMetadata datasetMetadata, List<ByteArrayInputStream> records) {
+    validateDatasetMetadataAndRecords(datasetMetadata, records);
+
+    final Set<Record> processedRecords = processRecordsAndRemoveDuplicates(datasetMetadata, records);
+
+    return new Dataset(datasetMetadata.getDatasetId(), processedRecords, records.size() - processedRecords.size());
+  }
+
+  private void validateDatasetMetadataAndRecords(DatasetMetadata datasetMetadata, List<ByteArrayInputStream> records) {
+    requireNonNull(datasetMetadata.getDatasetName(), "Dataset name must not be null");
+    requireNonNull(datasetMetadata.getCountry(), "Country must not be null");
+    requireNonNull(datasetMetadata.getLanguage(), "Language must not be null");
     requireNonNull(records, "Records must not be null");
     checkArgument(!records.isEmpty(), "Records must not be empty");
+  }
 
-    var recordObjectList = records.stream()
+  private Set<Record> processRecordsAndRemoveDuplicates(DatasetMetadata datasetMetadata, List<ByteArrayInputStream> records) {
+    return records
+        .stream()
         .map(ByteArrayInputStream::readAllBytes)
-        .map(record -> {
-          var recordId = xmlRecordProcessorService.getRecordId(record);
-          return Record.builder()
-              .recordId(recordId)
-              .europeanaId(EuropeanaIdCreator.constructEuropeanaIdString(recordId, datasetId))
-              .datasetId(datasetId)
-              .datasetName(name)
-              .country(country)
-              .language(language)
-              .content(record)
-              .build();
-        })
+        .map(recordItem -> getOptionalRecordFromProcessorService(datasetMetadata, recordItem))
+        .flatMap(Optional::stream)
         .collect(toSet());
+  }
 
-    // recordObjectList will remove duplicate records based on Record equals
-    return new Dataset(datasetId, recordObjectList, records.size() - recordObjectList.size());
+  private Optional<Record> getOptionalRecordFromProcessorService(DatasetMetadata datasetMetadata, final byte[] recordData) {
+    try {
+      final String recordId = xmlRecordProcessorService.getRecordId(recordData);
+      return Optional.of(Record.builder()
+                               .recordId(recordId)
+                               .europeanaId(EuropeanaIdCreator.constructEuropeanaIdString(recordId, datasetMetadata.getDatasetId()))
+                               .datasetId(datasetMetadata.getDatasetId())
+                               .datasetName(datasetMetadata.getDatasetName())
+                               .country(datasetMetadata.getCountry())
+                               .language(datasetMetadata.getLanguage())
+                               .content(recordData)
+                               .build());
+    } catch (IllegalArgumentException | RecordParsingException processorServiceException) {
```

Comment:
We should be logging the exception here if we are not throwing.

Reply:
yikes! correct, adding log.